searx/network/network.py

```diff
@@ -1,427 +1,505 @@
-        logger_name=None,
-    ):
-
-        self.enable_http = enable_http
-        self.verify = verify
-        self.enable_http2 = enable_http2
-        self.max_connections = max_connections
-        self.max_keepalive_connections = max_keepalive_connections
-        self.keepalive_expiry = keepalive_expiry
-        self.proxies = proxies
-        self.using_tor_proxy = using_tor_proxy
-        self.local_addresses = local_addresses
-        self.retries = retries
-        self.retry_on_http_error = retry_on_http_error
-        self.max_redirects = max_redirects
-        self._local_addresses_cycle = self.get_ipaddress_cycle()
-        self._proxies_cycle = self.get_proxy_cycles()
-        self._clients = {}
-        self._logger = logger.getChild(logger_name) if logger_name else logger
-        self.check_parameters()
-
-    def check_parameters(self):
-        for address in self.iter_ipaddresses():
-            if '/' in address:
-                ipaddress.ip_network(address, False)
-            else:
-                ipaddress.ip_address(address)
+    def __init__(self, settings: NetworkSettings):
+        """Creates a Network from a NetworkSettings"""
+        self._settings = settings
+        self._local_addresses_cycle = self._get_local_addresses_cycle()
+        self._proxies_cycle = self._get_proxy_cycles()
+        self._clients: Dict[Tuple, HTTPClient] = {}
+        self._logger = logger.getChild(settings.logger_name) if settings.logger_name else logger
 
-        if self.proxies is not None and not isinstance(self.proxies, (str, dict)):
-            raise ValueError('proxies type has to be str, dict or None')
+    @staticmethod
+    def from_dict(**kwargs):
+        """Creates a Network from a keys/values"""
+        return Network(NetwortSettingsDecoder.from_dict(kwargs))
+
+    def close(self):
+        """Close all the ABCHTTPClient hold by the Network"""
+        for client in self._clients.values():
+            client.close()
+
+    def check_configuration(self) -> bool:
+        """Check if the network configuration is valid.
+
+        Typical use case: check if the proxy is really a Tor proxy"""
+        try:
+            self._get_http_client()
+            return True
+        except Exception:  # pylint: disable=broad-except
+            self._logger.exception('Error')
+            return False
 
-    def iter_ipaddresses(self):
-        local_addresses = self.local_addresses
-        if not local_addresses:
-            return
-        if isinstance(local_addresses, str):
-            local_addresses = [local_addresses]
-        for address in local_addresses:
-            yield address
+    def get_context(self, timeout: Optional[float] = None, start_time: Optional[float] = None) -> NetworkContext:
+        """Return a new NetworkContext"""
+        context_cls = self._settings.retry_strategy.value
+        return context_cls(self._settings.retries, self._get_http_client, start_time, timeout)
```

Comment:
@dalf: Here we pass the arguments `self._settings.retries, self._get_http_client,` to the `RetryStrategy` (aka `NetworkContext`).  Is there any special reason why we not pass the `Network` object (here it is `self`) to the `NetworkContext` constructor.

When we pass the `NetworkContext` as argument, the function prototypes becomes less verbose and it scales better in future, when we might want to add more features to the context.

IMO it might be good, we pass the network object to Context & Clients .. so we can implement features in Context and Clients without the need to touch code like this one here to pas additional data from network to (e.g.) context.

Or do I oversee something?

Reply:
The purpose is to avoid circular import:
* searx.network.network imports searx.network.context
* if searx.network.context imports searx.network.network is going to be messy.

Reply:
Since the objects are already here, I guess you mean we have a circular import when we add type annotations?

Reply:
> Since the objects are already here, I guess you mean we have a circular import when we add type annotations?

Yes (that's the usage I can think of).

Note: it is also easier to write unit test.

Reply:
> Yes (that's the usage I can think of).

Haven't tested, but I think we can use [forward references](https://peps.python.org/pep-0484/#forward-references) .. where types are noted as strings:

```python

def foo(network: 'searx.network.network.Network'):
    pass
```
However, type-hints should not prevent us from writing the code we need.

> Note: it is also easier to write unit test.

I will have a look .. again: However, unit tests should not prevent us from writing the code we need :)

Thanks for your feedback, at least now I know that there are no hard reasons not to pass the network ... I think this simplifies the code a lot and scales better for future requirements.

Reply:
> I think this simplifies the code a lot

how?

The two values are copied:
https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/context.py#L81-L83

* `self._retries` is not a constant, so this attribute can't be skipped. See for example:
https://github.com/searxng/searxng/blob/ffcff586939a548c6cd9b279e2d0badea55a6656/searx/network/context.py#L260
* `http_client_factory` is wrapped. May be you have other implementation in mind to manage timeout?

The only thing is when we need additional parameter from the Network: I see that as a way to review the coupling between the two classes ; instead of calling any fields and methods from a Network instance and end up to entangle of two classes.

Reply:
Have a use case in mind ([the bing issue](https://github.com/return42/searxng/commit/ceba946e2547dabe56335a64cbd6229d731aeab9) / don't cache client, don't reuse SSL context in a bing-network), where I see its hard to pass an setting from the Network to the NetworkContext and from there down to the HTTPClient .. all these objects are working in one context, that is dominated by the Network object ..

Give me a try .. may I come to the conclusion its a bad idea, but for now I think all objects should have access the the central Network (.. to read settings from the config and decide what to do).